server/src/main/java/org/candlepin/controller/ProductManager.java

```diff
@@ -605,9 +606,47 @@ private Product applyProductChanges(Product entity, ProductDTO update, Owner own
             }
         }
 
+        this.applyProvidedProductChanges(entity, update, owner);
         return this.applyProductChanges(entity, update, contentMap);
     }
 
+    /**
+     * Applies the changes related to provided products from the given DTO to the specified entity.
+     *
+     * @param entity The entity to modify
+     * @param update The DTO containing the modifications to apply
+     * @param owner  An owner to use for resolving entity references
+     * @return The updated product entity
+     */
+    private Product applyProvidedProductChanges(Product entity, ProductDTO update, Owner owner) {
+
+        if (update.getProvidedProducts() != null) {
+            if (update.getProvidedProducts().isEmpty()) {
+                entity.setProvidedProducts(Collections.emptySet());
+            }
+            else {
+                Set<Product> products = new HashSet<>();
+
+                for (ProductDTO providedProductDTO : update.getProvidedProducts()) {
+                    if (providedProductDTO != null) {
+                        Product newProd = this.ownerProductCurator.getProductById(owner,
+                            providedProductDTO.getId());
+
+                        if (newProd == null) {
+                            log.debug("Provided product not found {}", providedProductDTO.getId());
+                        }
+
+                        products.add(newProd);
+                    }
+                }
+
+                entity.setProvidedProducts(products);
```

Comment:
Refactored bit of code. I think it is necessary to set it to entity directly as it involve product version upgrade during update process.

Reply:
You can still do the loop (and should, as you noted), but just skip building a set. At the point where the looked up product is added to the set, the ```addProvidedProduct``` method can be used instead. This would add it to the product's internal set all the same. Combine this with the initial clearing of the set (which needs to be done in both cases) and you can cover both cases without an explicit branch.

Reply:
Also, I just noticed there's no validation on the provided product ref. We should probably do some check to ensure the product isn't invalid and the lookup succeeded, similarly to how we process ProductContent in ```applyProductChanges```.

Reply:
Okay, using `addProvidedProduct` with initial clearing of the set. Added another validation on provided product reference.